vendors/honeycomb/src/OpenTelemetry/Vendor/Honeycomb.hs

```diff
@@ -0,0 +1,329 @@
+  deriving stock (Show, Eq)
+  deriving newtype (IsString)
+
+
+{- | Environment name in the Environments & Services data model (referred to as
+ \"Current\" in this package).
+
+ See https://docs.honeycomb.io/honeycomb-classic/ for more details.
+-}
+newtype EnvironmentName = EnvironmentName {unEnvironmentName :: Text}
+  deriving stock (Show, Eq)
+  deriving newtype (IsString)
+
+
+{- | Gets the Honeycomb configuration from the environment.
+
+    This does not do any HTTP.
+
+ FIXME(jadel): This should ideally fetch this from the tracer provider, but
+ it's nonobvious how to architect being able to do that (requires changes in
+ hs-opentelemetry-api). For now let's take a Tracer such that we
+ can fix it later, then do it the obvious way.
+-}
+getConfigPartsFromEnv :: (MonadIO m) => TracerProvider -> m (Maybe (Text, DatasetName))
+getConfigPartsFromEnv _ = do
+  mheaders <- liftIO $ lookupEnv "OTEL_EXPORTER_OTLP_HEADERS"
+  pure $ getValues =<< mheaders
+  where
+    discardLeft (Left _) = Nothing
+    discardLeft (Right a) = Just a
+
+    getValues headers = do
+      baggage <- discardLeft $ Baggage.decodeBaggageHeader (BS8.pack headers)
+      token <- Baggage.value <$> (HM.lookup headerHoneycombApiKey $ Baggage.values baggage)
+      let dataset = maybe "" Baggage.value (HM.lookup headerHoneycombLegacyDataset $ Baggage.values baggage)
+      pure (token, DatasetName dataset)
+
+
+{- | Gets the team name and environment name for the OTLP exporter using the API
+ key from the environment.
+
+ This calls Honeycomb.
+
+ N.B. Use 'Config.config' to construct a config from 'getConfigPartsFromEnv'.
+
+ N.B. The EnvironmentName will be Nothing if the API key is for a Honeycomb
+ Classic instance.
+-}
+getHoneycombData :: MonadIO m => Config.Config -> m (HoneycombTeam, Maybe EnvironmentName)
+getHoneycombData cfg = do
+  auth <- runReaderT Auth.getAuth cfg
+  let envSlug = Auth.slug . Auth.environment $ auth
+      mEnvSlug = if T.null envSlug then Nothing else Just (EnvironmentName envSlug)
+
+      team = HoneycombTeam . Auth.slug . Auth.team $ auth
+  pure (team, mEnvSlug)
+
+
+{- | Takes a 'Config.Config' and pokes around both Honeycomb HTTP API and the
+ trace environment to figure out where events will land in Honeycomb.
+-}
+resolveHoneycombTarget :: (MonadIO m) => TracerProvider -> Config.Config -> m (Maybe HoneycombTarget)
+resolveHoneycombTarget tracer cfg = do
+  (team, mEnvName) <- getHoneycombData cfg
+  let resources = getMaterializedResourcesAttributes . getTracerProviderResources $ tracer
+  pure $
+    HoneycombTarget team <$> case mEnvName of
+      -- There is an env name -> Current-Honeycomb
+      Just envName -> do
+        AttributeValue (TextAttribute serviceName) <- lookupAttribute resources "service.name"
+        pure $ Current envName (DatasetName serviceName)
+      -- Honeycomb Classic
+      Nothing -> do
+        pure $ Classic (Config.defaultDataset cfg)
```

Comment:
Just checking, will it default to classic if an error occurs? I don't remember the code well enough on the `honeycomb` package side.

Reply:
I'm not entirely sure what you mean. If we are not hardcoding the team name, it's always required to make an HTTP request to Honeycomb to figure out where the data is going regardless of Classic or Current. In the Classic case, it is required to get the team name and in Current, we need both team name and environment (which we can't get away with hardcoding for even in proprietary software).

Classic is distinguished by the auth response giving an empty string in the environment field.